numpy/random/generator.pyx

```diff
@@ -675,36 +674,40 @@ cdef class Generator:
                 size_i = size
                 pop_size_i = pop_size
                 # This is a heuristic tuning. should be improvable
-                if pop_size_i > 200 and (size > 200 or size > (10 * pop_size // size)):
+                if pop_size_i > 10000 and (size_i > (pop_size_i // 50)):
```

Comment:
How sensitive is the perf now with the improved shuffle? Here the full array is created whenever size is > 2% pop_size.  This could be memory unfriendly when pop_size is quite large (100_000_000) but size is still modest (~10%).

Reply:
Here are some timings on my machine:

original shuffle
```
In [4]: %timeit rg.choice(100_000_000, 2_000_000, replace=: False)                                            
235 ms ± 12.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
766 ms ± 9.58 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
7.06 s ± 85.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

specialized shuffle
```
In [3]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
187 ms ± 817 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
326 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
550 ms ± 7.94 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
4.89 s ± 16.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Floyd no shuffle
```
In [4]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
111 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [5]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
320 ms ± 6.46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
641 ms ± 9.64 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
5.64 s ± 90.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Floyd with shuffle
```
In [3]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
169 ms ± 836 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
496 ms ± 8.94 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
1.02 s ± 8.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
11.1 s ± 28.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
I sized the heuristic so that the speed is roughly the same when it switches from shuffle to Floyd. If we don't do a shuffle at the end, then the breakeven point is probably pop_size//20.

Reply:
This is purely timing based. You raise a good point about memory, Floyd is more memory efficient indeed, but it's a tradeoff.